pkg/middleware/batch.go

```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/go-chi/render"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// BatchResposeItem holds the structure for each item
+type BatchResposeItem struct {
+	Status       int         `json:"status"`
+	RequestID    string      `json:"requestID"`
+	Method       string      `json:"method"`
+	URL          string      `json:"url"`
+	OrigResponse interface{} `json:"origResponse"`
+}
+
+// BatchResponse has the structure for the final response
+type BatchResponse struct {
+	StartedAt  time.Time          `json:"startedAt"`
+	EndedAt    time.Time          `json:"endedAt"`
+	ErrorCount int                `json:"errorCount"`
+	Response   []BatchResposeItem `json:"response"`
+}
+
+// BatchWriter implements http.ResponseWriter
+type BatchWriter struct {
+	statusCode int
+	headerMap  http.Header
+	method     string
+	url        string
+
+	BResponse BatchResponse
+}
+
+// WriteHeader sets the status code
+func (rec *BatchWriter) WriteHeader(code int) {
```

Comment:
Is this method only used in tests?

Reply:
you need to define this method to overwrite the original WriteHeader()